### PR TITLE
Change Tensor summary to print tensor without losing shape information in output

### PR DESCRIPTION
**Pull request by guotong1988**
issue description https://github.com/tensorflow/tensorflow/issues/1702
Please close [this](https://github.com/tensorflow/tensorflow/pull/4884) because I make a github mistake there.
Please tell me , if there exists some code format error.

### CODE REVIEW

File: tensorflow/core/framework/tensor.cc [hunk 1]
```diff
@@ -677,14 +677,42 @@ bool Tensor::CanUseDMA() const {
...
+    strings::StrAppend(result, "[");
+    //as for each element, print the sub-element/right-dim
+    PrintOneDim(data, dim_index + 1, shape, limit,
+                data_index, shape_size, result);
+    strings::StrAppend(result, "]");
+  }
+}
+
+template <typename T>
+string SummarizeArray(int64 limit, int64 num_elts, const char* data,
```
Comment (drpngx) [inline tensorflow/core/framework/tensor.cc]: Prefer const TensorShape&. Also, out parameters (data) should go at the end.

File: tensorflow/core/framework/tensor.cc [hunk 2]
```diff
@@ -677,14 +677,42 @@ bool Tensor::CanUseDMA() const {
 #undef CASE
 
 namespace {
+//from left dim to right dim , recursive print
```
Comment (drpngx) [inline tensorflow/core/framework/tensor.cc]: Here and below, please abide to google C++ style guide, to wit:
Need one space after '//'
Comments should be a full sentence, including punctuation, for instance "Loop every element of one dim."

File: tensorflow/core/framework/tensor.cc [hunk 3]
```diff
@@ -677,14 +677,42 @@ bool Tensor::CanUseDMA() const {
...
 template <typename T>
-string SummarizeArray(int64 limit, int64 num_elts, const char* data) {
+void PrintOneDim(T *data, int dim_index, gtl::InlinedVector<int64, 4> shape,
```
Comment (drpngx) [inline tensorflow/core/framework/tensor.cc]: See below for parameter order:
dim_indx, shape, limit, shape_size, data, data_index, result.

File: tensorflow/core/framework/tensor.cc [hunk 4]
```diff
@@ -677,14 +677,42 @@ bool Tensor::CanUseDMA() const {
...
+                      TensorShape tensor_shape) {
   string ret;
   const T* array = reinterpret_cast<const T*>(data);
-  for (int64 i = 0; i < limit; ++i) {
-    if (i > 0) strings::StrAppend(&ret, " ");
-    strings::StrAppend(&ret, array[i]);
-  }
+
+  int64 data_index = 0;
+  gtl::InlinedVector<int64, 4> shape = tensor_shape.dim_sizes();
```
Comment (drpngx) [inline tensorflow/core/framework/tensor.cc]: Not strictly required by our style guide, but prefer to use const gtl::InlinedVector... shape.

File: tensorflow/core/framework/tensor.cc [hunk 5]
```diff
@@ -677,14 +677,42 @@ bool Tensor::CanUseDMA() const {
...
+  int shape_size = tensor_shape.dims();
```
Comment (drpngx) [inline tensorflow/core/framework/tensor.cc]: As above, const int shape_size

**Review by drpngx — CHANGES_REQUESTED** [review]

**Comment by guotong1988** [comment]
I finish.

File: tensorflow/core/framework/tensor.cc [hunk 6]
```diff
@@ -677,14 +677,43 @@ bool Tensor::CanUseDMA() const {
...
+    strings::StrAppend(result, "[");
+    // As for each element, print the sub-dim
+    PrintOneDim(dim_index + 1, shape, limit, shape_size,
+                data, data_index, result);
+    strings::StrAppend(result, "]");
+  }
+}
+
+template <typename T>
+string SummarizeArray(int64 limit, int64 num_elts, const TensorShape &tensor_shape,
```
Comment (drpngx) [inline tensorflow/core/framework/tensor.cc]: we use const TensorShape* tensor_shape

File: tensorflow/core/framework/tensor.cc [hunk 7]
```diff
@@ -677,14 +677,43 @@ bool Tensor::CanUseDMA() const {
 #undef CASE
 
 namespace {
+// Recursive print from left dim to right dim
 template <typename T>
-string SummarizeArray(int64 limit, int64 num_elts, const char* data) {
+void PrintOneDim(int dim_index, gtl::InlinedVector<int64, 4> shape, int64 limit, 
+                 int shape_size, T *data, int64 *data_index, string *result) {
```
Comment (drpngx) [inline tensorflow/core/framework/tensor.cc]: use T* data, int64* data_index, string* result

File: tensorflow/core/framework/tensor.cc [hunk 8]
```diff
@@ -677,14 +677,43 @@ bool Tensor::CanUseDMA() const {
 #undef CASE
 
 namespace {
+// Recursive print from left dim to right dim
```
Comment (drpngx) [inline tensorflow/core/framework/tensor.cc]: add punctuation + "Print ... recursively." is probably better phrasing, or Print recursively, from....

**Review by drpngx — CHANGES_REQUESTED** [review]
minor style nits

**Comment by guotong1988** [comment]
done

File: tensorflow/core/framework/tensor.cc [hunk 9]
```diff
@@ -703,14 +703,14 @@ void PrintOneDim(int dim_index, gtl::InlinedVector<int64, 4> shape, int64 limit,
 }
 
 template <typename T>
-string SummarizeArray(int64 limit, int64 num_elts, const TensorShape &tensor_shape,
```
Comment (drpngx) [inline tensorflow/core/framework/tensor.cc]: Sorry, it's fine to use a const ref, just collate the ampersand to the type, not the variable name.

File: tensorflow/core/framework/tensor.cc [hunk 10]
```diff
@@ -677,10 +677,10 @@ bool Tensor::CanUseDMA() const {
 #undef CASE
 
 namespace {
-// Recursive print from left dim to right dim
+// Print recursively from left dim to right dim
```
Comment (drpngx) [inline tensorflow/core/framework/tensor.cc]: Please put punctuation at the end of the comment ("." at the end), as per style guide.

**Comment by guotong1988** [comment]
finish

**Comment by guotong1988** [comment]
please wait a moment..

**Comment by guotong1988** [comment]
done

File: tensorflow/core/framework/tensor.cc [hunk 11]
```diff
@@ -677,14 +677,50 @@ bool Tensor::CanUseDMA() const {
 #undef CASE
 
 namespace {
+// Print from left dim to right dim recursively.
 template <typename T>
-string SummarizeArray(int64 limit, int64 num_elts, const char* data) {
+void PrintOneDim(int dim_index, gtl::InlinedVector<int64, 4> shape, int64 limit, 
+                 int shape_size, T* data, int64* data_index, string* result) {
+  if (*data_index >= limit) return;  
+  int element_count = shape[dim_index];
```
Comment (vrv) [inline tensorflow/core/framework/tensor.cc]: Use int64 here, shape is a vector of int64

File: tensorflow/core/framework/tensor.cc [hunk 12]
```diff
@@ -677,14 +677,50 @@ bool Tensor::CanUseDMA() const {
...
+  // Reach the rightmost dim of the tensor.
```
Comment (vrv) [inline tensorflow/core/framework/tensor.cc]: // We have reached the right-most dimension of the tensor.

**Comment by guotong1988** [comment]
Done.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Review by drpngx — APPROVED** [review]

**Comment by drpngx** [comment]
Please take a look at: https://ci.tensorflow.org/job/tensorflow-pull-requests-cpu-python3/2051/consoleFull

**Comment by guotong1988** [comment]
OK

**Comment by guotong1988** [comment]
I know the reason of tensor_test.cc . But I don't know the reason of TIMEOUT.

**Comment by guotong1988** [comment]
I know the reason of tensor_test.cc . But I don't know the reason of TIMEOUT.

**Comment by drpngx** [comment]
Github suffered an outage, and it likely affected jenkins as well, perhaps a repeat of the DDoS.

Jenkins, test this please.

**Comment by drpngx** [comment]
Something wrong with the tests, let's try tomorrow.

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
I think something about this change is indeed causing timeouts.  Possibly the new summarization code is too slow and is executed in some of the tests that are timing out.

**Comment by guotong1988** [comment]
So what should I do. Do you think it is a good idea to write two methods. One for print and another is the origin one.

**Comment by vrv** [comment]
I would try to investigate whether this change is indeed causing runtime of those tests to increase (you can test locally to find out), and then see if there's something you can do about it?  I can't say which is better without knowing more information from your end.

**Comment by guotong1988** [comment]
Oh.. Would you please tell me the instructions to run all the test?
I only know how to run a single one:)

**Comment by drpngx** [comment]
You need to run `bazel test -c opt tensorflow/...`, or select a test, like `//tensorflow/contrib/bayesflow:entropy_test` which took 300 seconds instead a just a few.

**Comment by guotong1988** [comment]
Oh..I 'm sorry. I think I can do it. It is 3 am now and I 'm a little slow with th brain.

**Comment by guotong1988** [comment]
I found the reason and fixed it.

**Comment by vrv** [comment]
Nice!  Can you add a test to tensor_test.cc that would have caught the problem too?

File: tensorflow/core/framework/tensor.cc [hunk 13]
```diff
@@ -717,6 +717,7 @@ string SummarizeArray(int64 limit, int64 num_elts, const TensorShape& tensor_sha
 
   int64 data_index = 0;
   const gtl::InlinedVector<int64, 4> shape = tensor_shape.dim_sizes();
+  if (shape.size() == 0) return "";
```
Comment (drpngx) [inline tensorflow/core/framework/tensor.cc]: Prefer the idiomatic if (shape.empty()) return "";

**Comment by guotong1988** [comment]
Done.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by guotong1988** [comment]
well I will fix it.

**Comment by guotong1988** [comment]
Next time I will run all tests before commit.

**Comment by guotong1988** [comment]
When the `shape` is empty , the `DebugString` method of the unit-test need the value to be return.

**Comment by drpngx** [comment]
Jenkins, test this please.

File: tensorflow/core/framework/tensor.cc [hunk 14]
```diff
@@ -677,14 +677,58 @@ bool Tensor::CanUseDMA() const {
...
+template <typename T>
+string SummarizeArray(int64 limit, int64 num_elts, const TensorShape& tensor_shape, 
+                      const char* data) {
   string ret;
   const T* array = reinterpret_cast<const T*>(data);
-  for (int64 i = 0; i < limit; ++i) {
-    if (i > 0) strings::StrAppend(&ret, " ");
-    strings::StrAppend(&ret, array[i]);
+
+  int64 data_index = 0;
```
Comment (drpngx) [inline tensorflow/core/framework/tensor.cc]: Move this after if (shape.empty()) {...}

Thanks!

**Comment by guotong1988** [comment]
Done.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by guotong1988** [comment]
I can't fix it because I don't know the reason.

**Comment by drpngx** [comment]
Probably a transient error in the config. Trying again.

Jenkins, test this please.

**Comment by guotong1988** [comment]
I have run `bazel test -c opt //tensorflow/contrib/learn:dnn_linear_combined_test` locally , but it passed.
It used 116.2s.

**Comment by drpngx** [comment]
Yeah, that test appears to have become flaky recently (failed last night's build). It seems to be a fairly recent problem.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Pull request merged**